contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1324 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, DepsMut, Env,
+    MessageInfo, QuerierWrapper, Response, StdResult, Timestamp, Uint128, WasmMsg,
+};
+use cw20::{Cw20ExecuteMsg, Cw20ReceiveMsg, Denom};
+use std::cmp::min;
+
+use crate::error::ContractError;
+use crate::error::ContractError::RewardsNotStarted;
+use crate::msg::{
+    ClaimableRewardsResponse, ExecuteMsg, InfoResponse, InstantiateMsg, QueryMsg, ReceiveMsg,
+};
+use crate::state::{Config, LastClaim, CONFIG, LAST_CLAIMED};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    validate_instantiate_msg(&env, &msg)?;
+    let config = Config {
+        start_block: msg.start_block,
+        end_block: msg.end_block,
+        payment_per_block: msg.payment_per_block,
+        total_amount: msg.total_amount,
+        denom: msg.denom.clone(),
+        staking_contract: deps.api.addr_validate(&*msg.distribution_token)?,
+        funded: match msg.denom {
+            Denom::Native(_) => true,
+            Denom::Cw20(_) => false,
+        },
+        payment_block_delta: msg.payment_block_delta,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default())
+}
+
+pub fn validate_instantiate_msg(env: &Env, msg: &InstantiateMsg) -> Result<(), ContractError> {
+    if env.block.height > msg.start_block {
+        return Err(ContractError::StartBlockAlreadyOccurred {});
+    }
+    if msg.start_block > msg.end_block {
+        return Err(ContractError::StartBlockAfterEndBlock {});
+    }
+    if msg.start_block % msg.payment_block_delta != 0 {
+        return Err(ContractError::StartBlockNotDivisibleByPaymentDelta {});
+    }
+    if msg.end_block % msg.payment_block_delta != 0 {
+        return Err(ContractError::EndBlockNotDivisibleByPaymentDelta {});
+    }
```

Comment:
Seems like this is what we actually want here. Slightly less restrictive.

(Note: invented a new error type here. Probably don't actually accept the suggestion.)

```suggestion
    if (msg.end_block - msg.start_block) % msg.payment_block_delta != 0 {
        return Err(ContractError::RewardsDurationNotDivisibleByPaymentDelta {});
    }
```

Reply:
yes, good suggestion